Add RootModules to api.GenerateServiceResponse

r-hang opened the pull request:
Even if a Thrift file has no service declaration and therefore no
req.RootServices entries, it's stil useful to have Module information
for the Thrift Files that ThriftRW is invoked with.

To provide an example of where this Module information can be
useful, the Directory field of a Module can help place ThriftRW
plugin generated code for Thrift files that don't contain service
declarations next to where ThriftRW has already generated its code
for that module.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thriftrw/thriftrw-go?pullRequest=449) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thriftrw/thriftrw-go?pullRequest=449) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thriftrw/thriftrw-go?pullRequest=449) it.</sub>

abhinav commented on gen/plugin_test.go:
```diff
@@ -9,13 +9,33 @@ import (
+	assert.Len(t, g.rootModules, 1)
+}
+
 func TestAddRootService(t *testing.T) {
```
should we rename this test?

r-hang replied:
Renamed to include RootModulesandServices to reflect additional functionality.

abhinav commented on gen/plugin_test.go:
```diff
@@ -171,9 +195,14 @@ func TestAddRootService(t *testing.T) {
 		g := newGenerateServiceBuilder(importer)
```
this is some old code; while we're here, would you mind making use of subtests here with t.Run?

```
for _, tt := range tests {
  t.Run(tt.desc, func(t *testing.T) {
    // ...
  })
}
```

r-hang replied:
Done

abhinav reviewed (commented):
Should we do a more end-to-end test as well? We can call [`gen.Generate`](https://pkg.go.dev/go.uber.org/thriftrw/gen?tab=doc#Generate) on a testdata Thrift file that imports other files, with a [fake/mock plugin](https://pkg.go.dev/go.uber.org/thriftrw/gen?tab=doc#Options.Plugin), and test the GenerateServiceRequest received by it.

r-hang commented:
Closing PR to turn around comments and restructure code.

r-hang commented:
@abhinav, yes! I just added a more 'end2end-ish test within the structure gen/generate_test.go! The existing setup hand creates a compile.Module instead of parsing one directly from a test Thrift, but where a compile.Module comes from is not central to this added functionality so I don't feel parsing a test Thrift is a hard constraint.

I mock the plugin Generate interface and I've added a custom matcher to verify that the GenerateServiceResponse does contain the new RootModules field that i've added.

abhinav commented on gen/generate_test.go:
```diff
@@ -395,3 +422,58 @@ func TestThriftPackageImporter(t *testing.T) {
+		return false
+	}
+
+	var intsA []int
+	for _, mID := range s.req.RootModules {
+		intsA = append(intsA, int(mID))
+	}
+
+	var intsB []int
```
if we're not matching other fields of GenerateServivceRequest, perhaps this should be `RootModulesMatcher`?

In fact, it can probably be `type rootModulesMatcher []api.ModuleID`.

abhinav commented on gen/plugin.go:
```diff
@@ -44,13 +44,17 @@ type generateServiceBuilder struct {
 func newGenerateServiceBuilder(i thriftPackageImporter) *generateServiceBuilder {
 	return &generateServiceBuilder{
 		GenerateServiceRequest: api.GenerateServiceRequest{
+			RootModules:   make([]api.ModuleID, 0),
```
we should give it a default capacity of 1 since that's what'll usually go in there.

```suggestion
			RootModules:   make([]api.ModuleID, 0, 1),
```

abhinav reviewed (approved):
don't forget the changelog
